src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/SetupComponentRegistrationService.cs

```diff
@@ -0,0 +1,259 @@
+    {
+        Requires.Argument(projectGuid != Guid.Empty, nameof(projectGuid), "Cannot be an empty GUID.");
+
+        await InitializeAsync(cancellationToken);
+
+        // Add an empty snapshot for the project for now.
+        ImmutableInterlocked.Update(
+            ref _snapshotByProjectGuid,
+            static (dic, projectGuid) => dic.Add(projectGuid, null),
+            projectGuid);
+
+        return new DisposableDelegate(() =>
+        {
+            ImmutableInterlocked.Update(
+                ref _snapshotByProjectGuid,
+                static (dic, projectGuid) => dic.Remove(projectGuid),
+                projectGuid);
+        });
+    }
+
+    public void SetProjectComponentSnapshot(Guid projectGuid, UnconfiguredSetupComponentSnapshot snapshot)
+    {
+        Requires.Argument(projectGuid != Guid.Empty, nameof(projectGuid), "Cannot be an empty GUID.");
+        Requires.Argument(_snapshotByProjectGuid.ContainsKey(projectGuid), nameof(projectGuid), "Project GUID must be registered.");
+
+        if (ImmutableInterlocked.Update(ref _snapshotByProjectGuid, static (dic, pair) => dic.SetItem(pair.Key, pair.Value), (Key: projectGuid, Value: snapshot)))
+        {
+            TryPublish();
+        }
+    }
+
+    private void TryPublish()
+    {
+        // Grab a reference to this now so that we can work with a consistent view of the solution data without locking.
+        ImmutableDictionary<Guid, UnconfiguredSetupComponentSnapshot?> snapshotByProjectGuid = _snapshotByProjectGuid;
+
+        if (!HaveAllSnapshots())
+        {
+            return;
+        }
+
+        Assumes.Present(_setupCompositionService);
+        Assumes.Present(_missingComponentRegistrationService);
+
+        Dictionary<Guid, IReadOnlyCollection<string>>? missingComponentsByProjectGuid = null;
+
+        foreach ((Guid projectGuid, UnconfiguredSetupComponentSnapshot? snapshot) in snapshotByProjectGuid)
+        {
+            Assumes.NotNull(snapshot);
+
+            List<string>? missingComponents = null;
+
+            foreach (string componentId in snapshot.ComponentIds)
+            {
+                if (IsMissingComponent(componentId))
+                {
+                    missingComponents ??= new();
+                    missingComponents.Add(componentId);
+                }
+            }
+
+            if (missingComponents is not null)
+            {
+                missingComponentsByProjectGuid ??= new();
+                missingComponentsByProjectGuid.Add(projectGuid, missingComponents);
+            }
+        }
+
+        if (missingComponentsByProjectGuid is not null)
+        {
+            Task task = _missingComponentRegistrationService.RegisterMissingComponentsAsync(missingComponentsByProjectGuid, cancellationToken: default);
+
+            _projectFaultHandlerService.Forget(task, project: null, ProjectFaultSeverity.Recoverable);
+        }
+
+        return;
+
+        bool HaveAllSnapshots()
```

Comment:
nit: you might consider inlining the usage of this function - the return false/true isn't necessary if we just return at callsite. for me, it hinders readability

Reply:
I like how this function is self-documenting. The method's already quite long, and there's zero runtime overhead of pulling this precondition out into a local function.